jibDockerBuild fails to pull source image from gcr

<!-- Thanks for filing an issue! Please use this template to provide more context for the issue . -->

**Description of the issue**: Unable to build Docker image from custom base image hosted in gcr with gradle.plugin.com.google.cloud.tools:jib-gradle-plugin:0.9.10 

I've got Docker on the path: PATH=/Applications/Docker.app/Contents/Resources/bin:...

Already have the image locally after executing:
- gcloud auth configure-docker
- gcloud auth print-access-token | docker login -u oauth2accesstoken --password-stdin https://gcr.io
- docker pull gcr.io/ir-devops-playground/oracle-jdk1.8:alpine3.8 

**Environment**: 
- Mac El Capitan 10.11.6
- Docker version 18.06.1-ce, build e68fc7a

**`jib-gradle-plugin` Configuration**:
Both my & service IAM account have "Storage Object Viewer" role. Tried 2 authentication approaches:

```
jib {
        from {
            image = "gcr.io/ir-devops-playground/oracle-jdk1.8:alpine3.8"
            auth {
                username = '_json_key'
                password =  file("${project.rootDir}/ir-devops-playground-33d34617c277.json").text
            }
        }
  }
```
&
```
jib {
        from {
            image = "gcr.io/ir-devops-playground/oracle-jdk1.8:alpine3.8"
            auth {
                username = 'oauth2accesstoken'
                password = 'gcloud auth print-access-token'.execute().text.trim()
            }
        }
  }
```

**Log output**:
```
./gradlew :impactradius-account:jibDockerBuild --stacktrace

Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':impactradius-account:jibDockerBuild'.

Caused by: org.gradle.api.GradleException: Build to Docker daemon failed
        at com.google.cloud.tools.jib.gradle.BuildDockerTask.buildDocker(BuildDockerTask.java:118)
       ....
Caused by: java.io.IOException: 'docker load' command failed with output: 
        at com.google.cloud.tools.jib.docker.DockerClient.load(DockerClient.java:110)
        at com.google.cloud.tools.jib.builder.steps.LoadDockerStep.afterPushBaseImageLayerFuturesFuture(LoadDockerStep.java:92)
        at com.google.common.util.concurrent.CombinedFuture$CallableInterruptibleTask.runInterruptibly(CombinedFuture.java:181)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
```


**Additional Information**: 
Originally tried pulling the image from our insecure Nexus Docker repo with `allowInsecureRegistries=true` but when that didn't work, so I pushed the image up to gcr.


<!-- Thanks for contributing! -->

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheMetadata.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheMetadata.java
@@ -34,7 +34,7 @@ class CacheMetadata {
   static class Builder {
 
     private final ImageLayers.Builder<CachedLayerWithMetadata> layersBuilder =
-        ImageLayers.builder();
+        ImageLayers.<CachedLayerWithMetadata>builder().removeDuplicates();
 
     private Builder(ImageLayers<CachedLayerWithMetadata> initialLayers) {
       layersBuilder.addAll(initialLayers);

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageLayers.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageLayers.java
@@ -79,16 +79,15 @@ public class ImageLayers<T extends Layer> implements Iterable<T> {
     }
 
     public ImageLayers<T> build() {
-      ImmutableList<T> layers;
-      if (removeDuplicates) {
-        // keep last occurrence of each layer by adding in reverse order,
-        // and then reversing the result
-        Set<T> dedupedButReversed = new LinkedHashSet<T>(Lists.reverse(this.layers));
-        layers = ImmutableList.copyOf(dedupedButReversed).reverse();
-      } else {
-        layers = ImmutableList.copyOf(this.layers);
+      if (!removeDuplicates) {
+        return new ImageLayers<>(ImmutableList.copyOf(layers), layerDigestsBuilder.build());
       }
-      return new ImageLayers<>(layers, layerDigestsBuilder.build());
+
+      // LinkedHashSet maintains the order but keeps the first occurrence. Keep last occurrence by
+      // adding elements in reverse, and then reversing the result
+      Set<T> dedupedButReversed = new LinkedHashSet<T>(Lists.reverse(this.layers));
+      ImmutableList<T> deduped = ImmutableList.copyOf(dedupedButReversed).reverse();
+      return new ImageLayers<>(deduped, layerDigestsBuilder.build());
     }
   }
 

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageLayers.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageLayers.java
@@ -18,18 +18,23 @@ package com.google.cloud.tools.jib.image;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
 import javax.annotation.Nullable;
 
-/** Holds the layers for an image. Makes sure that there are no duplicate layers. */
+/** Holds the layers for an image. */
 public class ImageLayers<T extends Layer> implements Iterable<T> {
 
   public static class Builder<T extends Layer> {
 
-    private final LinkedHashSet<T> layers = new LinkedHashSet<>();
+    private final List<T> layers = new ArrayList<>();
     private final ImmutableSet.Builder<DescriptorDigest> layerDigestsBuilder =
         ImmutableSet.builder();
+    private boolean removeDuplicates = false;
 
     /**
      * Adds a layer. Removes any prior occurrences of the same layer.
@@ -43,11 +48,7 @@ public class ImageLayers<T extends Layer> implements Iterable<T> {
      */
     public Builder<T> add(T layer) throws LayerPropertyNotFoundException {
       layerDigestsBuilder.add(layer.getBlobDescriptor().getDigest());
-
-      // Remove necessary to move layer to the end of the LinkedHashSet.
-      layers.remove(layer);
       layers.add(layer);
-
       return this;
     }
 
@@ -67,8 +68,27 @@ public class ImageLayers<T extends Layer> implements Iterable<T> {
       return this;
     }
 
+    /**
+     * Remove any duplicate layers, keeping the last occurrence of the layer.
+     *
+     * @return this
+     */
+    public Builder<T> removeDuplicates() {
+      removeDuplicates = true;
+      return this;
+    }
+
     public ImageLayers<T> build() {
-      return new ImageLayers<>(ImmutableList.copyOf(layers), layerDigestsBuilder.build());
+      ImmutableList<T> layers;
+      if (removeDuplicates) {
+        // keep last occurrence of each layer by adding in reverse order,
+        // and then reversing the result
+        Set<T> dedupedButReversed = new LinkedHashSet<T>(Lists.reverse(this.layers));
+        layers = ImmutableList.copyOf(dedupedButReversed).reverse();
+      } else {
+        layers = ImmutableList.copyOf(this.layers);
+      }
+      return new ImageLayers<>(layers, layerDigestsBuilder.build());
     }
   }
 

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/image/ImageLayersTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/image/ImageLayersTest.java
@@ -74,12 +74,37 @@ public class ImageLayersTest {
   }
 
   @Test
-  public void testAddLayer_sameAsLastLayer() throws LayerPropertyNotFoundException {
+  public void testAddLayer_maintainDuplicates() throws LayerPropertyNotFoundException {
+    // must maintain duplicate
+    List<Layer> expectedLayers =
+        Arrays.asList(
+            mockCachedLayer,
+            mockReferenceLayer,
+            mockDigestOnlyLayer,
+            mockUnwrittenLayer,
+            mockCachedLayer);
+
+    ImageLayers<Layer> imageLayers =
+        ImageLayers.builder()
+            .add(mockCachedLayer)
+            .add(mockReferenceLayer)
+            .add(mockDigestOnlyLayer)
+            .add(mockUnwrittenLayer)
+            .add(mockCachedLayer)
+            .build();
+
+    Assert.assertEquals(expectedLayers, imageLayers.getLayers());
+  }
+
+  @Test
+  public void testAddLayer_removeDuplicates() throws LayerPropertyNotFoundException {
+    // remove duplicates: last layer should be kept
     List<Layer> expectedLayers =
         Arrays.asList(mockReferenceLayer, mockDigestOnlyLayer, mockUnwrittenLayer, mockCachedLayer);
 
     ImageLayers<Layer> imageLayers =
         ImageLayers.builder()
+            .removeDuplicates()
             .add(mockCachedLayer)
             .add(mockReferenceLayer)
             .add(mockDigestOnlyLayer)

--- a/jib-gradle-plugin/CHANGELOG.md
+++ b/jib-gradle-plugin/CHANGELOG.md
@@ -12,6 +12,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Keep duplicate layers to match container history ([#1017](https://github.com/GoogleContainerTools/jib/pull/1017))
+
 ## 0.9.10
 
 ### Added

--- a/jib-maven-plugin/CHANGELOG.md
+++ b/jib-maven-plugin/CHANGELOG.md
@@ -13,6 +13,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Keep duplicate layers to match container history ([#1017](https://github.com/GoogleContainerTools/jib/pull/1017))
+
 ## 0.9.10
 
 ### Added

